### PR TITLE
feat(js): generate state, code verifier and code challenge

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

- Add generators
    - generateState
    - generateCodeVerifier
    - generateCodeChallenge

## Linear Issue Reference

LOG-1354

## Testing

- generateState:
    - result should be random value
- generateCodeVerifier
    - result should be random value
- generateCodeChallenge
    - should generate different codeChallenge with different code verifiers
    - should generate same codeChallenge with same code verifiers

---
@logto-io/eng

### CODE REVIEW

File: packages/js/src/utils/generators.test.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    const verifier2 = generateCodeVerifier();
+    expect(verifier1).not.toEqual(verifier2);
+  });
+});
+
+describe('generateCodeChallenge', () => {
+  test('with fixed length', async () => {
+    const verifier = generateCodeVerifier();
+    const challenge = await generateCodeChallenge(verifier);
+    expect(challenge.length).toEqual(Math.ceil(32 * 1.34));
```
Comment (simeng-li) [inline packages/js/src/utils/generators.test.ts:52]: nit: could have a comment for these magic numbers.

File: packages/js/src/utils/generators.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Generates random bytes for state and encodes them in url safe base64
+ */
+export const generateState = () => generateRandomString(CODE_VERIFIER_MAX_LENGTH);
+
+/**
+ * Generates random bytes for nonce and encodes them in url safe base64
+ */
+export const generateNonce = () => generateRandomString(CODE_VERIFIER_MAX_LENGTH);
```
Comment (simeng-li) [inline packages/js/src/utils/generators.ts:39]: Do we cover this method?  Looks like missing from the conversion doc

**Review by simeng-li — COMMENTED** [review]
please check

File: packages/js/src/utils/generators.ts (hunk 2 above)
Comment (IceHe) [reply packages/js/src/utils/generators.ts:39]: I'm good with pre-defining unused utility function for now.
(add it according to the implementation of `oidc-client`)

@xiaoyijun any ideas?

- A. remove for now
- B. reserve it

Comment (xiaoyijun) [reply packages/js/src/utils/generators.ts:39]: A

File: packages/js/src/utils/generators.test.ts (hunk 1 above)
Comment (IceHe) [reply packages/js/src/utils/generators.test.ts:41]: Added, PTAL.

**Comment by IceHe** [comment]
Removed the dependency of `nanoid`

File: packages/js/src/utils/generators.test.ts (hunk 1 above)
Comment (gao-sun) [reply packages/js/src/utils/generators.test.ts:41]: this line is not expected. we do not care the length of the generated string

File: packages/js/src/utils/generators.test.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  generateCodeChallenge,
+  generateCodeVerifier,
+  generateState,
+} from './generators';
+
+describe('generateState', () => {
+  test('with fixed length', () => {
+    const state = generateState();
+    expect(state.length).toEqual(CODE_VERIFIER_MAX_LENGTH);
+  });
```
Comment (gao-sun) [inline packages/js/src/utils/generators.test.ts:12]: all these `with fixed length` s are not expected

File: packages/js/src/utils/generators.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import { UrlSafeBase64 } from '@silverhand/essentials';
+import { fromUint8Array } from 'js-base64';
+
+/**
+ * Length Range of Code Verifier
+ *
+ * The length of code verifier ranges from 43 to 128.
+ * @link [Client Creates a Code Verifier](https://datatracker.ietf.org/doc/html/rfc7636#section-4.1)
+ */
+export const CODE_VERIFIER_MAX_LENGTH = 128;
```
Comment (gao-sun) [inline packages/js/src/utils/generators.ts:12]: why do we need this constant? our sdk convention specified we hard code all random string in a raw length of 64.

File: packages/js/src/utils/generators.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * The strings of length 96 will contain 128 characters after encoding in base64.
+ * Because the base64 encoding cause an overhead of 33-36%.
+ * @link [Base64](Reference: https://en.wikipedia.org/wiki/Base64)
+ */
+const CODE_VERIFIER_RAW_RANDOM_DATA_LENGTH = 96;
```
Comment (gao-sun) [inline packages/js/src/utils/generators.ts:19]: this doesn't fit our sdk convention..

File: packages/js/src/utils/generators.test.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+  test('with random value', () => {
+    const verifier1 = generateCodeVerifier();
+    const verifier2 = generateCodeVerifier();
+    expect(verifier1).not.toEqual(verifier2);
+  });
+});
+
+describe('generateCodeChallenge', () => {
```
Comment (gao-sun) [inline packages/js/src/utils/generators.test.ts:34]: you also need some static [test cases](https://github.com/logto-io/swift/blob/master/Tests/LogtoTests/LogtoUtilitiesTests.swift#L18-L38) to prove the algorithm is correct, actually that's the core

**Review by gao-sun — CHANGES_REQUESTED** [review]

File: packages/js/src/utils/generators.test.ts (hunk 3 above)
Comment (IceHe) [reply packages/js/src/utils/generators.test.ts]: Removed.

File: packages/js/src/utils/generators.ts (hunk 4 above)
Comment (IceHe) [reply packages/js/src/utils/generators.ts]: Removed.

File: packages/js/src/utils/generators.ts (hunk 5 above)
Comment (IceHe) [reply packages/js/src/utils/generators.ts]: Removed.

File: packages/js/src/utils/generators.test.ts (hunk 1 above)
Comment (IceHe) [reply packages/js/src/utils/generators.test.ts]: Removed.

File: packages/js/src/utils/generators.test.ts (hunk 6 above)
Comment (IceHe) [reply packages/js/src/utils/generators.test.ts]: Added.

File: packages/js/src/utils/generators.test.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+  test('dealing with same code verifier should be equal', async () => {
+    const codeVerifier =
+      'KH5wVipNo1W0cpKnx0csh_WlmY5fpmAiYl46OBBc2SE6j93RtKo8j7M230jQ9DF0IWBFl1j_UsrI_5dcpHdo1w';
+    const codeChallenge1 = await generateCodeChallenge(codeVerifier);
+    const codeChallenge2 = await generateCodeChallenge(codeVerifier);
+    expect(codeChallenge1).toEqual(codeChallenge2);
+  });
+
+  test('dealing with code verifier containing latin1 character', async () => {
```
Comment (IceHe) [inline packages/js/src/utils/generators.test.ts:66]: Discuss: is it necessary test cases?
(added according to [unit tests](https://github.com/logto-io/swift/blob/master/Tests/LogtoTests/LogtoUtilitiesTests.swift#L18-L38) from Swift SDK)

Comment (IceHe) [reply packages/js/src/utils/generators.test.ts:66]: @logto-io/eng any idea?

File: packages/js/src/utils/generators.test.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    const codeVerifier = 'Á';
+    const codeChallenge = await generateCodeChallenge(codeVerifier);
+    expect(codeChallenge).toEqual('p3yvZiKYauPicLIDZ0W1peDz4Z9KFC-9uxtDfoO1KOQ');
+  });
+
+  test('dealing with code verifier containing emoji character', async () => {
```
Comment (IceHe) [inline packages/js/src/utils/generators.test.ts:72]: ditto

File: packages/js/src/utils/generators.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import { fromUint8Array } from 'js-base64';
+
+const DEFAULT_RANDOM_DATA_LENGTH = 64;
+
+/**
+ * @param length The length of the raw random data.
+ */
+function generateRandomString(length: number) {
+  const randomValues = crypto.getRandomValues(new Uint8Array(length));
+  const base64String = fromUint8Array(randomValues);
```
Comment (IceHe) [inline packages/js/src/utils/generators.ts:13]: If use `crypto.getRandomValues(new Uint8Array(length))`, which returns `Uint8Array` data, to generate random data, how to implement base64-encode and base64-decode under both Node and browser-like environments?

( js/js should be able to run under both Node and browser-like environments)

- from `Uint8Array` to base64 string
- from base64 string to `Uint8Array`

Any idea? Ask for help @logto-io/eng

Comment (IceHe) [reply packages/js/src/utils/generators.ts:13]: image

File: packages/js/src/utils/generators.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * Calculates the S256 PKCE code challenge for an arbitrary code verifier and encodes it in url safe base64
+ *
+ * @param {String} codeVerifier Code verifier to calculate the S256 code challenge for
+ * @link [Client Creates the Code Challenge](https://datatracker.ietf.org/doc/html/rfc7636#section-4.2)
+ */
+export const generateCodeChallenge = async (codeVerifier: string): Promise<string> => {
+  const encoded = new TextEncoder().encode(codeVerifier);
+  const challenge = new Uint8Array(await crypto.subtle.digest('SHA-256', encoded));
+  return UrlSafeBase64.replaceNonUrlSafeCharacters(fromUint8Array(challenge));
```
Comment (simeng-li) [inline packages/js/src/utils/generators.ts:38]: Looks like `fromUint8Array` takes a second param which returns a URI safe  Base64

Comment (IceHe) [reply packages/js/src/utils/generators.ts]: Fixed, PTAL.

Comment (IceHe) [reply packages/js/src/utils/generators.ts]: Again.

File: packages/js/src/utils/generators.test.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
+import { toUint8Array } from 'js-base64';
+
+import { generateCodeChallenge, generateCodeVerifier, generateState } from './generators';
+
+const isUrlSafe = (input: string) => /^[\w-]*$/.test(input);
```
Comment (gao-sun) [inline packages/js/src/utils/generators.test.ts:5]: we can add this into `essentials`

File: packages/js/src/utils/generators.test.ts (hunk 6 above)
Comment (gao-sun) [reply packages/js/src/utils/generators.test.ts:43]: i didn't mean 'dealing with different code verifiers should not be equal' and 'dealing with same code verifier should be equal' should use static strings.. 😂 should add a new `test()` to test static values, please see the link above

File: packages/js/src/utils/generators.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
+/** @link [Proof Key for Code Exchange by OAuth Public Clients](https://datatracker.ietf.org/doc/html/rfc7636) */
+
+import { fromUint8Array } from 'js-base64';
+
+const DEFAULT_RANDOM_DATA_LENGTH = 64;
+
+/**
+ * @param length The length of the raw random data.
+ */
+function generateRandomString(length: number) {
```
Comment (gao-sun) [inline packages/js/src/utils/generators.ts:10]:
```suggestion
const generateRandomString = (length = DEFAULT_RANDOM_DATA_LENGTH) =>
```

Comment (gao-sun) [reply packages/js/src/utils/generators.ts:10]: also we do not have the convention to use UPPER_CASE for constants in js. camelCase would be fine.

File: packages/js/src/utils/generators.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Calculates the S256 PKCE code challenge for an arbitrary code verifier and encodes it in url safe base64
+ *
+ * @param {String} codeVerifier Code verifier to calculate the S256 code challenge for
+ * @link [Client Creates the Code Challenge](https://datatracker.ietf.org/doc/html/rfc7636#section-4.2)
+ */
+export const generateCodeChallenge = async (codeVerifier: string): Promise<string> => {
+  const encodedCodeVerifier = new TextEncoder().encode(codeVerifier);
+  const codeChallenge = new Uint8Array(await crypto.subtle.digest('SHA-256', encodedCodeVerifier));
+  return fromUint8Array(codeChallenge, true);
```
Comment (gao-sun) [inline packages/js/src/utils/generators.ts:35]: we can remove related func in `essentials` if needed

File: packages/js/src/utils/generators.test.ts (hunk 11 above)
Comment (gao-sun) [reply packages/js/src/utils/generators.test.ts:5]: no hurries, you can add a linear issue and do it later

File: packages/js/src/utils/generators.ts (hunk 13 above)
Comment (gao-sun) [reply packages/js/src/utils/generators.ts:35]: no hurries too

File: packages/js/src/utils/generators.test.ts (hunk 11 above)
Comment (IceHe) [reply packages/js/src/utils/generators.test.ts:5]: use `UrlSafeBase64` in #128 , you can take a look

File: packages/js/src/utils/generators.ts (hunk 13 above)
Comment (IceHe) [reply packages/js/src/utils/generators.ts:35]: ditto: use UrlSafeBase64 in #128 , you can take a look

File: packages/js/src/utils/generators.ts (hunk 12 above)
Comment (IceHe) [reply packages/js/src/utils/generators.ts]: Fixed, PTAL.

File: packages/js/src/utils/generators.test.ts (hunk 6 above)
Comment (IceHe) [reply packages/js/src/utils/generators.test.ts:46]: Fixed, PTAL.

File: packages/js/src/utils/generators.test.ts (hunk 11 above)
Comment (gao-sun) [reply packages/js/src/utils/generators.test.ts:8]: ok

**Review by gao-sun — APPROVED** [review]
lgtm

File: packages/js/src/utils/generators.test.ts (hunk 11 above)
Comment (IceHe) [reply packages/js/src/utils/generators.test.ts:8]: Added issue LOG-1378 and will do it later

**Pull request merged**